TailLight/device.cpp

```diff
@@ -1,46 +1,83 @@
 #include "driver.h"
 #include <Hidport.h>
+#include "SetBlack.h"
 
 EVT_WDF_IO_QUEUE_IO_DEVICE_CONTROL EvtIoDeviceControlFilter;
 
 
-VOID EvtSetBlackTimer(_In_ WDFTIMER  Timer) {
-    KdPrint(("TailLight: EvtSetBlackTimer begin\n"));
+NTSTATUS PnpNotifyDeviceInterfaceChange(
+    _In_ PVOID pvNotificationStructure,
+    _Inout_opt_ PVOID pvContext)
+/*++
+Routine Description:
+    Filters out HID class device interface arrivals.
 
-    WDFDEVICE device = (WDFDEVICE)WdfTimerGetParentObject(Timer);
-    NT_ASSERTMSG("EvtSetBlackTimer device NULL\n", device);
+Arguments:
+    PVOID - One of several possible notification structures. All we care about
+            for this implementation is DEVICE_INTERFACE_CHANGE_NOTIFICATION.
 
-    NTSTATUS status = SetFeatureColor(device, 0);
-    if (!NT_SUCCESS(status)) {
-        KdPrint(("TailLight: EvtSetBlackTimer failure NTSTATUS=0x%x\n", status));
-        return;
+    PVOID - The WDFDEVICE that we received from EvtDriverDeviceAdd.
+
+--*/
+
+{
+
+    //KdPrint(("TailLight: PnpNotifyDeviceInterfaceChange enter\n"));
+    NT_ASSERTMSG("WDFDEVICE not passed in!", pvContext);
+
+    if (pvNotificationStructure == NULL) {
+        return STATUS_SUCCESS;
     }
 
-    KdPrint(("TailLight: EvtSetBlackTimer end\n"));
+    PDEVICE_INTERFACE_CHANGE_NOTIFICATION pDevInterface =
+        (PDEVICE_INTERFACE_CHANGE_NOTIFICATION)pvNotificationStructure;
+
+    ASSERT(IsEqualGUID(pDevInterface->InterfaceClassGuid,
+        GUID_DEVINTERFACE_HID));
+
+    {     
+        // Assumption: Device will arrive before removal.
+        if (IsEqualGUID(pDevInterface->Event,
+            GUID_DEVICE_INTERFACE_ARRIVAL)) {
+
+            // Opening a device may trigger PnP operations. Ensure that either a
+            // timer or a work item is used when opening up a device.
+            // Refer to page 356 of "Programming The Microsoft Windows Driver
+            // Model", 2nd edition by Walter Oney and IoGetDeviceObjectPointer.
+            // 
+            // Also note that eventually several system threads will be running
+            // in parallel. As a PnP system thread must not block, serializing
+            // would require a queue.
+            //
+            // NOTE: It is possible for us to get blocked waiting for a system
+            // thread. One solution would be to use a timer that spawns a
+            // system thread on timeout vs. a wait loop.
+            // Always must return success, so ignore result.
```

Comment:
Please clarify what is meant by "Always must return success" here. Are you referring to `CreateWorkItemForIoTargetOpenDevice` or `PnpNotifyDeviceInterfaceChange`? If the latter, then where is the "must return success" requirement documented?